File: tests/test_normalization.py
import pandas as pd
import pytest

from crispr_analysis_utils import counts_per_million


def test_counts_per_million_normalizes_columns():
    counts = pd.DataFrame(
        {"sample_a": [100, 300], "sample_b": [50, 50]},
        index=["guide_1", "guide_2"],
    )

    result = counts_per_million(counts)

    assert result.loc["guide_1", "sample_a"] == pytest.approx(250_000)
    assert result.loc["guide_2", "sample_a"] == pytest.approx(750_000)
    assert result["sample_a"].sum() == pytest.approx(1_000_000)
    assert result["sample_b"].sum() == pytest.approx(1_000_000)


def test_counts_per_million_rejects_zero_totals():
    counts = pd.DataFrame({"sample_a": [0, 0]})

    with pytest.raises(ValueError, match="total count zero"):
        counts_per_million(counts)

File: tests/testthat/test-normalization.R
test_that("counts_per_million normalizes columns", {
  counts <- matrix(
    c(100, 300, 50, 50),
    nrow = 2,
    dimnames = list(c("guide_1", "guide_2"), c("sample_a", "sample_b"))
  )

  result <- counts_per_million(counts)

  expect_equal(result["guide_1", "sample_a"], 250000)
  expect_equal(result["guide_2", "sample_a"], 750000)
  expect_equal(colSums(result), c(sample_a = 1e6, sample_b = 1e6))
})

test_that("counts_per_million rejects zero totals", {
  counts <- matrix(c(0, 0), ncol = 1)

  expect_error(counts_per_million(counts), "total count zero")
})
